### PR TITLE
Add meta registration for _foreach_norm (2nd try)

### PR DESCRIPTION
**Pull request by janeyx99**
The first try reused TensorListMetadata, which caused illegal memory access issues when there were too many tensors in the list. We just launch multiple kernels with a simpler version of the struct (to minimize kernels launched).

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #119927

### CODE REVIEW

File: aten/src/ATen/native/cuda/ForeachReduceOp.cu [hunk 1]
```diff
@@ -200,11 +223,28 @@ std::vector<Tensor> foreach_tensor_norm_cuda(
...
+            const size_t num_tensors_this_kernel =
+                (i < num_kernels - 1 || ntensors % MAX_TENSORS_PER_KERNEL == 0)
+                ? MAX_TENSORS_PER_KERNEL
+                : (ntensors % MAX_TENSORS_PER_KERNEL);
+
+            TensorListAddresses addr_struct;
+            for (auto j = 0; j < num_tensors_this_kernel; j++) {
+              addr_struct.addresses[j] = vec_res[i * MAX_TENSORS_PER_KERNEL + j]
+                                             .mutable_data_ptr<scalar_t>();
+            }
```
Comment (janeyx99) [inline aten/src/ATen/native/cuda/ForeachReduceOp.cu:238]: Not sure if there is a way to abstract this common code to its own function due to needing to know what `scalar_t` is

File: aten/src/ATen/native/cuda/ForeachReduceOp.cu [hunk 2]
```diff
@@ -178,9 +196,14 @@ std::vector<Tensor> foreach_tensor_norm_cuda(
   }
   const auto options = tensors[0].options();
   auto output_per_tensor = at::zeros(
-      {ntensors * max_chunks_per_tensor},
+      {static_cast<long>(ntensors) * max_chunks_per_tensor},
```
Comment (albanD) [inline aten/src/ATen/native/cuda/ForeachReduceOp.cu:199]:
```suggestion
      {static_cast<uint64_t>(ntensors) * max_chunks_per_tensor},
```
?

File: test/test_foreach.py [hunk 3]
```diff
@@ -646,6 +646,20 @@ def test_foreach_l2_large_value_input(self, device, dtype, op):
                 for i, e in enumerate(expect)))
         self.assertEqual(expect, actual, equal_nan=False)
 
+    @onlyCUDA
+    @ops(foreach_reduce_op_db, allowed_dtypes=floating_types())
+    def test_foreach_norm_big_num_tensors(self, device, dtype, op):
```
Comment (albanD) [inline test/test_foreach.py:651]: Why is this only running foreach_norm give that it is decorated with "foreach_reduce_op_db"?

Comment (janeyx99) [reply test/test_foreach.py:651]: foreach_norm is our only reduce op but yes i see your point + will remove!

File: aten/src/ATen/native/cuda/ForeachReduceOp.cu (hunk 2 above)
Comment (janeyx99) [reply aten/src/ATen/native/cuda/ForeachReduceOp.cu:199]: ah

**Review by albanD — APPROVED** [review]
SGTM!

File: aten/src/ATen/native/cuda/ForeachReduceOp.cu [hunk 4]
```diff
@@ -223,11 +263,28 @@ std::vector<Tensor> foreach_tensor_norm_cuda(
...
-              ret_per_tensor.mutable_data_ptr<scalar_t>(),
-              max_chunks_per_tensor);
-          C10_CUDA_KERNEL_LAUNCH_CHECK();
+
+          const size_t num_kernels = ceil_div(ntensors, MAX_TENSORS_PER_KERNEL);
+          for (auto i = 0; i < num_kernels; i++) {
+            const size_t num_tensors_this_kernel =
+                (i < num_kernels - 1 || ntensors % MAX_TENSORS_PER_KERNEL == 0)
+                ? MAX_TENSORS_PER_KERNEL
+                : (ntensors % MAX_TENSORS_PER_KERNEL);
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/ForeachReduceOp.cu:272]:
```
const size_t num_tensors_this_kernel = std::min(MAX_TENSORS_PER_KERNEL, ntensors - i * MAX_TENSORS_PER_KERNEL)
```

File: aten/src/ATen/native/cuda/ForeachReduceOp.cu (hunk 1 above)
Comment (ngimel) [reply aten/src/ATen/native/cuda/ForeachReduceOp.cu:238]: you could have a function templated on NormType that would call `AT_DISPATCH_FLOATING_TYPES_AND2`

**Review by ngimel — COMMENTED** [review]
looks good, minor nits

**Pull request closed without merging**